core/src/consensus.rs

```diff
@@ -319,6 +336,91 @@ impl Tower {
+                    //    or any of its descendants
+                    if !descendants.is_empty()
+                        || candidate_slot == last_vote
+                        || ancestors.get(&candidate_slot).unwrap().contains(last_vote)
+                    {
+                        continue;
+                    }
+
+                    // By the time we reach here, any ancestors of the `last_vote`,
+                    // should have been filtered out, as they all have a descendant,
+                    // namely the `last_vote` itself.
+                    assert!(!last_vote_ancestors.contains(candidate_slot));
+
+                    // Evaluate which vote accounts in the bank are locked out
+                    // in the interval candidate_slot..last_vote, which means
+                    // finding any lockout intervals in the `lockout_intervals` tree
+                    // for this bank that contain `last_vote`.
+                    let lockout_intervals = &progress
+                        .get(&candidate_slot)
+                        .unwrap()
+                        .fork_stats
+                        .lockout_intervals;
+                    // Find any locked out intervals in this bank with endpoint >= last_vote,
+                    // implies they are locked out at last_vote
+                    for (_, value) in lockout_intervals.range((Included(last_vote), Unbounded)) {
+                        for (lockout_interval_start, vote_account_pubkey) in value {
+                            // Only count lockouts on slots that are:
+                            // 1) Not ancestors of `last_vote`
+                            // 2) Not from before the current root as we can't determine if
+                            // anything before the root was an ancestor of `last_vote` or not
+                            if !last_vote_ancestors.contains(lockout_interval_start)
+                                // The check if the key exists in the ancestors map
+                                // is equivalent to checking if the key is above the
+                                // current root.
+                                && ancestors.contains_key(lockout_interval_start)
+                                && !locked_out_vote_accounts.contains(vote_account_pubkey)
+                            {
+                                let stake = epoch_vote_accounts
+                                    .get(vote_account_pubkey)
+                                    .map(|(stake, _)| *stake)
+                                    .unwrap_or(0);
+                                locked_out_stake += stake;
+                                locked_out_vote_accounts.insert(vote_account_pubkey);
+                            }
+                        }
+                    }
+                }
+                (locked_out_stake as f64 / total_stake as f64) > SWITCH_FORK_THRESHOLD
+            })
+            .unwrap_or(true)
```

Comment:
This doesn't generate the proof yet, want to make sure these incremental changes don't break anything first